core/drivers/stm32_rtc.c

```diff
@@ -0,0 +1,848 @@
+		   ((cal->dr & RTC_DR_YU_MASK) >> RTC_DR_YU_SHIFT) + 2000;
+}
+
+TEE_Result stm32_rtc_get_calendar(struct stm32_rtc_calendar *calendar)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!rtc_dev.pclk)
+		return TEE_ERROR_GENERIC;
+
+	res = clk_enable(rtc_dev.pclk);
+	if (res)
+		return res;
+
+	stm32_rtc_read_calendar(calendar);
+
+	/* RTC may need to be read twice, depending of clocks configuration */
+	if (rtc_dev.flags & RTC_FLAGS_READ_TWICE) {
+		uint32_t tr_save = calendar->tr;
+
+		stm32_rtc_read_calendar(calendar);
+
+		if (calendar->tr != tr_save)
+			stm32_rtc_read_calendar(calendar);
+	}
+
+	clk_disable(rtc_dev.pclk);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Get the subsecond value.
+ *
+ * Subsecond is a counter that grows from 0 to stm32_rtc_get_subsecond_scale()
+ * every second.
+ */
+static uint32_t stm32_rtc_get_subsecond(struct stm32_rtc_calendar *cal)
+{
+	uint32_t prediv_s = io_read32(get_base() + RTC_PRER) &
+			    RTC_PRER_PREDIV_S_MASK;
+	uint32_t ss = cal->ssr & RTC_SSR_SS_MASK;
+
+	return prediv_s - ss;
+}
+
+/*
+ * Get the subsecond scale.
+ *
+ * Number of subseconds in a second is linked to RTC PREDIV_S value.
+ * The more PREDIV_S will be high, the more subseconds will be precise.
+ */
+static uint32_t stm32_rtc_get_subsecond_scale(void)
+{
+	return (io_read32(get_base() + RTC_PRER) & RTC_PRER_PREDIV_S_MASK) + 1;
+}
+
+/* Return relative difference of ticks count between two rtc calendar */
+static signed long long
+stm32_rtc_diff_subs_tick(struct stm32_rtc_calendar *cur,
+			 struct stm32_rtc_calendar *ref,
+			 unsigned long tick_rate)
+{
+	return (((signed long long)stm32_rtc_get_subsecond(cur) -
+		 (signed long long)stm32_rtc_get_subsecond(ref)) *
+		(signed long long)tick_rate) /
+		stm32_rtc_get_subsecond_scale();
+}
+
+/* Return relative difference in milliseconds on subsecond */
+static signed long long stm32_rtc_diff_subs_ms(struct stm32_rtc_calendar *cur,
+					       struct stm32_rtc_calendar *ref)
+{
+	return stm32_rtc_diff_subs_tick(cur, ref, MS_PER_SEC);
+}
+
+/* Return absolute difference in milliseconds on seconds-in-day fraction */
+static signed long long stm32_rtc_diff_time_ms(struct optee_rtc_time *current,
+					       struct optee_rtc_time *ref)
+{
+	signed long long curr_s = 0;
+	signed long long ref_s = 0;
+
+	curr_s = (signed long long)current->tm_sec +
+		 (((signed long long)current->tm_min +
+		  (((signed long long)current->tm_hour * 60))) * 60);
+
+	ref_s = (signed long long)ref->tm_sec +
+		(((signed long long)ref->tm_min +
+		 (((signed long long)ref->tm_hour * 60))) * 60);
+
+	return (curr_s - ref_s) * 1000U;
+}
+
+/* Return absolute difference in milliseconds on day-in-year fraction */
+static signed long long stm32_rtc_diff_date_ms(struct optee_rtc_time *current,
+					       struct optee_rtc_time *ref)
+{
+	uint32_t diff_in_days = 0;
+	uint32_t m = 0;
+	const uint8_t month_len[] = {
+		31, 28, 31, 30, 31, 30,
+		31, 31, 30, 31, 30, 31
+	};
+
```

Comment:
I think this implementation does not well handle case where `current` relates to a date lower `ref` reference.
Maybe building a decimal value for both as done in `rtc_timecmp()` (commit "drivers: rtc: add time comparison in RTC API") could help comparing most of the time values to help rejecting absolute time difference computation when argument do no match this constraint.